### PR TITLE
Cleanup and refactoring, mostly prepared for custom values

### PR DESCRIPTION
**Pull request by IsaacElenbaas**
I've done my best to condense this to just one evaluate function and remove conditions. It fits more with AutoShift's behavior (record stuff on press, evaluate on release or another key down [and now timeout]) and is easier to wrap your head around.

Adding support for custom values should at this point be a 10-minute task, aside from somehow recording whether the shifted key was the one pressed (for if its release isn't immediately triggered and the shift state could have been changed). Simply releasing both the unshifted and shifted base keys is also an option.

### CODE REVIEW

**Comment by ceriinas** [comment]
As far as I can tell behavior is unchanged, but this does not rely on the key's own releases which makes custom values possible.

File: quantum/process_keycode/process_auto_shift.c [hunk 1]
```diff
@@ -107,113 +100,83 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
  * If the autoshift key is released before the delay has elapsed, register the
  * key without a shift.
- *
- * If another key is pressed, register the key.
  */
-static void autoshift_check_record(uint16_t keycode, keyrecord_t *record) {
-    if (!autoshift_flags.in_progress) {
-        return;
-    }
+static void autoshift_end(uint16_t keycode, uint16_t now, bool matrixTrigger) {
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c]: s/matrixTrigger/matrix_trigger/

File: quantum/process_keycode/process_auto_shift.c [hunk 2]
```diff
@@ -79,24 +59,37 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
+            return false;
+        }
+#        ifndef AUTO_SHIFT_NO_AUTO_REPEAT
     }
+#        endif
 #    endif
 
     // Record the keycode so we can simulate it later.
     autoshift_lastkey           = keycode;
+    autoshift_time = record->event.time;
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c]: Run `clang-format`

File: quantum/process_keycode/process_auto_shift.c [hunk 3]
```diff
@@ -107,113 +100,83 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
-            // The key is registered; flush any shift state for the
-            // non-autoshiftable key.
+
+#    if TAP_CODE_DELAY > 0
+                wait_ms(TAP_CODE_DELAY);
+#    endif
+        unregister_code(autoshift_lastkey);
+        autoshift_flush_shift();
+    }
+    else {
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c]: Should be on previous line according to https://beta.docs.qmk.fm/developing-qmk/c-development/coding_conventions_c

File: quantum/process_keycode/process_auto_shift.c (hunk 1 above)
Comment (ceriinas) [reply quantum/process_keycode/process_auto_shift.c]: Too much Python recently lol

File: quantum/process_keycode/process_auto_shift.c [hunk 4]
```diff
@@ -107,113 +100,83 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
+/** \brief Simulates auto-shifted key releases when timeout is hit
  *
  *  Can be called from \c matrix_scan_user so that auto-shifted keys are sent
  *  immediately after the timeout has expired, rather than waiting for the key
  *  to be released.
  */
 void autoshift_matrix_scan(void) {
-    if (!autoshift_flags.in_progress || autoshift_flags.registered) {
-        return;
+    const uint16_t now = timer_read();
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c]: `timer_read()` is an expensive operation (requires synchronization and such depending on the arch), and matrix_scan is called frequently.  Minimize the calls to `timer_read` by making them conditional on autoshift_flags as before.

File: quantum/process_keycode/process_auto_shift.c [hunk 5]
```diff
@@ -225,6 +188,12 @@ void set_autoshift_timeout(uint16_t timeout) { autoshift_timeout = timeout; }
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record) {
     if (record->event.pressed) {
+        if (autoshift_flags.in_progress) {
+            autoshift_end(KC_NO, record->event.time, false);
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c:183]: Excuse my ignorance, but can you explain under what circumstances KC_NO occurs here?

File: quantum/process_keycode/process_auto_shift.c [hunk 6]
```diff
@@ -107,113 +100,83 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
         // Time since the initial press was recorded.
-        const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
-        if (!autoshift_flags.registered && elapsed < autoshift_timeout) {
-            // Auto-shiftable key is being released before the shift timeout;
-            // simulate the original press then let the usual processing take
-            // care of the release.
-            register_code(keycode);
+        const uint16_t elapsed = TIMER_DIFF_16(now, autoshift_time);
+        if (elapsed < autoshift_timeout) {
+            // Change this and below to 16 if custom keys are added.
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c]: Remove comments alluding to custom keys for now

**Review by p00ya — CHANGES_REQUESTED** [review]
Thanks :)

I'll do some testing with your change as well.

File: quantum/process_keycode/process_auto_shift.c (hunk 5 above)
Comment (ceriinas) [reply quantum/process_keycode/process_auto_shift.c:183]: That line is to evaluate the current AutoShift key on another down. I used to have a `if (keycode == autoshift_lastkey)` in the `in_progress` section of `autoshift_end` (as testing if it was an AS key would still be true if it was one cancelling another), but as that's gone the `KC_NO` isn't necessary. It does make it clear the line isn't to test the current key, though - I'll add a comment too.

**Comment by ceriinas** [comment]
Pressing one autoshift key, pressing another, and releasing the first causes both to be unshifted regardless of the second one's hold time. Not sure how to fix here, but found a solution if custom shifts is merged.

File: quantum/process_keycode/process_auto_shift.c (hunk 5 above)
Comment (ceriinas) [reply quantum/process_keycode/process_auto_shift.c:183]: Still not required here but the fix for [this](https://github.com/p00ya/qmk_firmware/pull/1#issuecomment-666553392) in custom shift pull means everything in that slot is necessary.

**Comment by p00ya** [comment]
Issues from manual testing (with `AUTO_SHIFT_REPEAT`):

1. Both shift keys are getting held.  I can see that the code is only doing `add_weak_mods(MOD_BIT(KC_LSFT))`, so possibly that's something misleading in how `add_weak_mods` works.  Why this is bad: double-shift can have a special action in the OS (e.g. `setxkbmap -option grp:shifts_toggle` binds double-shift to changing the OS keymap in X11).
2. The shift keys are stuck on after the auto-shifted key is released.  Release the shift key when the last auto-shifted key is released (unless the user manually held down the left shift key themselves).  Why this is bad: something like "click and insert" in a document will instead "select and replace" because you're actually shift-clicking.

You'll have to rebase to get rid of the conflicts (sorry - I'm rebasing over qmk/develop myself).

> Pressing one autoshift key, pressing another, and releasing the first causes both to be unshifted regardless of the second one's hold time. Not sure how to fix here, but found a solution if custom shifts is merged.

FYI I couldn't repro the issue you mentioned (if I hold `1` until it repeats as `!`, then hold `2`, then release `1`, then `shift` is still correctly held for me).

**Comment by ceriinas** [comment]
Shoot, thought I had already fixed the conflicts. Oh well, will try again.

Tested with `xbindkeys --multikey`, only LSFT is being hit. Reason for 2 was `del_weak_mods` that doesn't trigger a keyboard report.

The issue with releasing is when under the tapping term - pressing the second key triggers evaluation of the first, and releasing the first key still in the second one's hold time causes the latter to be evaluated and released as non-shifted, even if you're still holding it. The cause is `autoshift_end` not having a specified key to release, and defaulting to the currently autoshifted one if it is in progress. The fix from custom shifts might actually work, I don't see why it wouldn't anymore. Will test.

File: quantum/process_keycode/process_auto_shift.c [hunk 7]
```diff
@@ -247,26 +207,28 @@ bool process_auto_shift(uint16_t keycode, keyrecord_t *record) {
...
 #    endif
+            if (record->event.pressed) {
                 return autoshift_press(keycode, record);
-
-            default:
-                break;
-        }
+            } else {
+                autoshift_end(keycode, record->event.time, false);
+                return false;
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c:229]: What would it take to rewrite this so that we could return true here (at least for the non-custom case)?

I tried to avoid eating records since the behaviour becomes less transparent to things downstream (like `process_record_user`).

File: quantum/process_keycode/process_auto_shift.c [hunk 8]
```diff
@@ -107,114 +92,81 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
+        wait_ms(TAP_CODE_DELAY);
+#    endif
+        unregister_code(autoshift_lastkey);
+        del_weak_mods(MOD_BIT(KC_LSFT));
+    } else {
+        // Release after keyrepeat.
+        unregister_code(keycode);
+        if (keycode == autoshift_lastkey) {
+            // This will only fire when the key was the last auto-shiftable
+            // pressed. That prevents aaaaBBBB then releasing a from unshifting
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c:134]: nit: do you mind putting quotes around the keys here ("releasing a from" is a bit hard to parse eagerly).

File: quantum/process_keycode/process_auto_shift.c [hunk 9]
```diff
@@ -79,24 +53,35 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
...
-        return true;
+#    ifdef AUTO_SHIFT_REPEAT
+    const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
+#        ifndef AUTO_SHIFT_NO_AUTO_REPEAT
+    if (!autoshift_flags.lastshifted) {
+#        endif
+        if (elapsed < TAPPING_TERM && keycode == autoshift_lastkey) {
+            // Allow a tap-then-hold for keyrepeat.
+            if (!autoshift_flags.lastshifted) {
+                register_code(autoshift_lastkey);
```
Comment (p00ya) [inline quantum/process_keycode/process_auto_shift.c:69]: Would be good if we could return true here rather than registering the key ourselves.  Though you'd miss updating autoshift_time below if you returned directly.

**Review by p00ya — CHANGES_REQUESTED** [review]

File: quantum/process_keycode/process_auto_shift.c (hunk 7 above)
Comment (ceriinas) [reply quantum/process_keycode/process_auto_shift.c:229]: Basically just not mergng this. It's near impossible to return true and still have control of keyrepeat to make this not a breaking change, and this refactor relies heavily on the assumption that it will be the one pressing and releasing keys (as I was preparing for custom ones). Your original pull is simpler if you want to go with only-keyrepeat, though I would probably submit another cleanup pull to make it handle shift more like this.

IMO that is more of a QMK-wide issue. We need a way to let other processes know that a key was pressed but that it's been handled and they shouldn't press anything, avoiding `return false;` everywhere is a pain.
